pulumi/src/pvcResolver.ts

```diff
@@ -0,0 +1,57 @@
+import * as k8s from '@pulumi/kubernetes'
+import { VolumeSnapshot } from './volumeSnapshotClient';
+
+interface PVCFromBackup extends k8s.types.input.core.v1.PersistentVolumeClaim {
+  datasource: {
+    name: string;
+    kind: string;
+    apiGroup: string;
+  }
+}
+
+export const getVolumeClaimTemplates = (serviceName: string, storageSize: string, snapshots?: VolumeSnapshot[]): Array<k8s.types.input.core.v1.PersistentVolumeClaim> => {
```

Comment:
alternate solution:
  - filter includes both checks
  - snapshots can be ordered by snapshot client to remove extra sort
  - use `PersistentVolumeClaim` class
  - spread in `dataSource` if there is a valid snapshot found
  - use snapshot contents for `dataSource`
```ts
export const getVolumeClaimTemplates = (
  serviceName: string,
  storageSize: string,
  snapshots: VolumeSnapshot[]
): Array<k8s.types.input.core.v1.PersistentVolumeClaim> => {
  const snapshot = snapshots.filter(snapshot => snapshot.metadata.name.includes(serviceName) && snapshot.status.readyToUse)[0]

  const pvc: k8s.types.input.core.v1.PersistentVolumeClaim = {
    metadata: {
      name: `data-${serviceName}`,
    },
    spec: {
      accessModes: ['ReadWriteOnce'],
      storageClassName: 'ebs-csi-gp2',
      resources: {
        requests: {
          storage: storageSize,
        },
      },
      ...(snapshot && {
        dataSource: {
          name: snapshot.metadata.name,
          kind: snapshot.kind,
          apiGroup: snapshot.apiVersion.split('/')[0],
        },
      }),
    },
  }

  return [pvc]
}
```

Reply:
Nice, much cleaner